src/main/java/net/shrimpworks/unreal/archive/mirror/MirrorClient.java

```diff
@@ -22,41 +23,61 @@
  * progress).
  */
 public class MirrorClient implements Consumer<MirrorClient.Downloader> {
-
-	private final Deque<Content> content;
+	private final int retryLimit = 4;
+	private Deque<Content> content;
+	private Deque<Content> retryQueue;
 	private final Path output;
 	private final int concurrency;
 	private final ExecutorService executor;
 
 	private final Progress progress;
 
 	private final long totalCount;
-	private final CountDownLatch counter;
+	private CountDownLatch counter;
```

Comment:
Because the `counter` is not longer `final` and is being re-assigned by the `mirrorThread`, but access by threads of the `executor` (in the `accept()` method), it should be `volatile` so the correct "current" counter can be referenced by multiple threads more safely.

Reply:
Ah, that went missing while I tried a different approach ;-)

Reply:
Cool; I don't think it needs to be `final` though, it does need to be `volatile`.

Having the value `volatile` ensures that all threads are reading the correct value after the reassignment happens; otherwise it's possible that some of the executor threads finish their work while the `mirrorThread` replaces the reference, and then there's the possibility of the executor thread(s) calling `countDown()` on a stale `counter` reference.

I hope that makes sense? :)